### ISSUE TITLE
[DistDGL] Random crash in DistDataloader with PyTorch > 1.12 

### ISSUE THREAD

**Issue by chang-l** [dmlc/dgl#5480]
This issue has been going on for a while.  
Since pytorch 1.13, DistDGL is observed with random crashes in `DistDataloader` when start moving/accessing shared memory concurrently (memory corruption). A simple repo could be DGL's unit test `tests/distributed/test_mp_dataloader.py` with PyTorch 1.13.0, 1.13.1 and 2.0+.

I think it is introduced by this PyTorch commit: https://github.com/pytorch/pytorch/commit/b25a1ce22d965a852da5979e7d6af9fe91518451, which releases the GIL before `storage_copy` that triggers the memory corruption. A detailed discussion of this issue can be found at this thread: https://github.com/pytorch/pytorch/issues/95606.

The fixes can be 
(1) Revert this commit https://github.com/pytorch/pytorch/commit/b25a1ce22d965a852da5979e7d6af9fe91518451; and rebuild PyTorch 1.13+.  It is a walkaround and I tested it locally.
(2) It already has a fix in upstream, here: https://github.com/pytorch/pytorch/commit/6ea790c5b600338f848ec240dbb7cfe248e6f2de. We can wait for it to be merged in later releases... 

## To Reproduce

```
python3 -m pytest -v --capture=tee-sys --junitxml=pytest_distributed.xml --durations=100 tests/distributed/test_mp_dataloader.py
```
The stack trace and error msg is:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/dgl/dgl-source/tests/distributed/test_mp_dataloader.py", line 457, in start_node_dataloader
    for idx, (_, _, blocks) in zip(
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/dist_dataloader.py", line 117, in __next__
    result = self._get_data_from_result_queue()
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/dist_dataloader.py", line 129, in _get_data_from_result_queue
    ret = self.pool.get_result(self.name, timeout=timeout)
  File "/usr/local/lib/python3.8/dist-packages/dgl/distributed/dist_context.py", line 180, in get_result
    dl_name, data = self.result_queue.get(timeout=timeout)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
  File "/usr/local/lib/python3.8/dist-packages/torch/multiprocessing/reductions.py", line 100, in rebuild_tensor
    t = torch._utils._rebuild_tensor(storage, storage_offset, size, stride)
  File "/usr/local/lib/python3.8/dist-packages/torch/_utils.py", line 148, in _rebuild_tensor
    return t.set_(storage._untyped_storage, storage_offset, size, stride)
RuntimeError: Trying to resize storage that is not resizable
```

## Environment

 - DGL Version (e.g., 1.0): 1.0
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 2.0
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed DGL (`conda`, `pip`, source): source
 - CUDA/cuDNN version (if applicable): 12.0
 
## Additional context

It also leads the application [here](https://github.com/awslabs/graphstorm) failed with PyTorch 1.13+. Again, reverting https://github.com/pytorch/pytorch/commit/b25a1ce22d965a852da5979e7d6af9fe91518451 can fix it temporarily.

**Comment by Rhett-Ying** [comment]
DGL CI is using pytorch 1.12.0, so it didn't capture this bug.

**Comment by jermainewang** [comment]
Look like there is nothing we can do on our side for the moment. We will try to incorporate more tests in the release pipeline to preemptively capture those issues in the future.

**Comment by Rhett-Ying** [comment]
The fix is supposed to be ready in latest `torch 2.0.1`?

**Comment by chang-l** [comment]
>The fix is supposed to be ready in latest torch 2.0.1?

Seems not there yet. There is no release tags under that commit (https://github.com/pytorch/pytorch/commit/6ea790c5b600338f848ec240dbb7cfe248e6f2de).

**Comment by CongWeilin** [comment]
Same issue here https://github.com/awslabs/graphstorm/issues/199#issuecomment-1598012610

**Comment by Rhett-Ying** [comment]
Quoted:

Hey!
It will be in the next release. You can find the timing for it here: https://dev-discuss.pytorch.org/t/pytorch-release-2-1-0/1271
You can also use nightly binaries if you need this feature earlier!

https://github.com/pytorch/pytorch/pull/96664#issuecomment-1622478776

**Comment by aishwaryyasarkar** [comment]
Hi, I'm experiencing the same issue and I've already installed the PyTorch nightly build. Is dgl nightly build necessary too to resolve the error?

**Comment by chang-l** [comment]
Now, this issue has been resolved in the latest PyTorch 2.1.0 release. https://github.com/pytorch/pytorch/commit/6ea790c5b600338f848ec240dbb7cfe248e6f2de


Closing
